rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+    rcl_action_goal_handle_t * goal_handle;
+    for (size_t i = 0; i < total_num_goals; ++i) {
+      goal_handle = action_server->impl->goal_handles[i];
+      rcl_ret_t ret = rcl_action_goal_handle_get_info(goal_handle, &goal_info);
+      assert(RCL_RET_OK == ret);
+
+      const int64_t goal_nanosec = _goal_info_stamp_to_nanosec(&goal_info);
+      if (rcl_action_goal_handle_is_cancelable(goal_handle) &&
+        ((goal_nanosec <= request_nanosec) || uuidcmp(request_uuid, goal_info.uuid)))
+      {
+        goal_handles_to_cancel[num_goals_to_cancel++] = goal_handle;
+      }
+    }
+  }
+
+  rcl_ret_t ret_final = RCL_RET_OK;
+  if (0u == num_goals_to_cancel) {
+    cancel_response->msg.goals_canceling.data = NULL;
+    cancel_response->msg.goals_canceling.size = 0u;
+    goto cleanup;
+  }
+
+  // Allocate space in response
+  rcl_ret_t ret = rcl_action_cancel_response_init(
+    cancel_response, num_goals_to_cancel, allocator);
+  if (RCL_RET_OK != ret) {
+    if (RCL_RET_BAD_ALLOC == ret) {
+      ret_final = RCL_RET_BAD_ALLOC;  // error already set
+    }
+    ret_final = RCL_RET_ERROR;  // error already set
+    goto cleanup;
+  }
+
+  // Transition goals to canceling and add to response
+  rcl_action_goal_handle_t * goal_handle;
+  for (size_t i = 0; i < num_goals_to_cancel; ++i) {
+    goal_handle = goal_handles_to_cancel[i];
+    ret = rcl_action_update_goal_state(goal_handle, GOAL_EVENT_CANCEL);
+    if (RCL_RET_OK == ret) {
+      ret = rcl_action_goal_handle_get_info(
+        goal_handle, &cancel_response->msg.goals_canceling.data[i]);
+    }
+    if (RCL_RET_OK != ret) {
+      ret_final = RCL_RET_ERROR;  // error already set
+    }
+  }
+cleanup:
+  allocator.deallocate(goal_handles_to_cancel, allocator.state);
+  return ret_final;
+}
+
+rcl_ret_t
+rcl_action_send_cancel_response(
+  const rcl_action_server_t * action_server,
+  void * ros_cancel_response)
+{
+  SEND_SERVICE_RESPONSE(cancel);
+}
+
+const char *
+rcl_action_server_get_action_name(const rcl_action_server_t * action_server)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return NULL;  // error already set
+  }
+  return action_server->impl->action_name;
+}
+
+const rcl_action_server_options_t *
+rcl_action_server_get_options(const rcl_action_server_t * action_server)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return NULL;  // error already set
+  }
+  return &action_server->impl->options;
+}
+
+rcl_ret_t
+rcl_action_server_get_goal_handles(
+  const rcl_action_server_t * action_server,
+  rcl_action_goal_handle_t *** goal_handles,
+  size_t * num_goals)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(goal_handles, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(num_goals, RCL_RET_INVALID_ARGUMENT);
+  *goal_handles = action_server->impl->goal_handles;
```

Comment:
@jacobperron should we check for `NULL == *goal_handles`?

Reply:
No. We can set `*goal_handles` to a valid pointer, regardless of it's initial value.